lib/ood_core/job/adapters/sge/batch.rb

```diff
@@ -0,0 +1,194 @@
+# Object used for simplified communication with a SGE batch server
+#
+# @api private
+class OodCore::Job::Adapters::Sge::Batch
+  using OodCore::Refinements::HashExtensions
+  require "ood_core/job/adapters/sge/qstat_xml_j_r_listener"
+  require "ood_core/job/adapters/sge/qstat_xml_r_listener"
+  require "ood_core/job/adapters/sge/helper"
+  require "ood_core/refinements/hash_extensions"
+  require "tempfile"
+  require 'time'
+
+  begin
+    require "ood_core/job/adapters/drmaa"
+    require 'singleton'
+
+    DRMMA_TO_OOD_STATE_MAP = {
+      DRMAA::STATE_UNDETERMINED          => :undetermined,
+      DRMAA::STATE_QUEUED_ACTIVE         => :queued,
+      DRMAA::STATE_SYSTEM_ON_HOLD        => :queued_held,
+      DRMAA::STATE_USER_ON_HOLD          => :queued_held,
+      DRMAA::STATE_USER_SYSTEM_ON_HOLD   => :queued_held,
+      DRMAA::STATE_RUNNING               => :running,
+      DRMAA::STATE_SYSTEM_SUSPENDED      => :suspended,
+      DRMAA::STATE_USER_SUSPENDED        => :suspended,
+      DRMAA::STATE_USER_SYSTEM_SUSPENDED => :suspended,
+      DRMAA::STATE_DONE                  => :completed,
+      DRMAA::STATE_FAILED                => :completed
+    }
+
+    # The one and only connection with DRMAA
+    # Attempting to instantiate a DRMAA::Session more than once causes it to crash
+    class SessionSingleton < DRMAA::Session
+      include Singleton
+    end
+  rescue LoadError; end
```

Comment:
Should we log a warning here? Or is it expected functionality that this load attempt happens multiple times in the normal execution?

If we want to log a warning we could open an issue to add a ood_core wide logger. There might be an issue already open. Something where an app can set the wide logger to its logger i.e. `OodCore.logger = Rails.logger`.

Reply:
Change to load in a `self.load` that is executed conditionally if sge_root is defined AND raise an exception if there is a LoadError or let LoadError bubble up.